### PR TITLE
Extend Advisor recommendation update to service groups

### PR DESCRIPTION
**Pull request by ravigupta99**
## What does this PR do?

Extends the existing `advisor recommendation update` tool from #3303 so the same MCP tool can update either subscription-scoped or Azure service-group-scoped Advisor recommendations.

The unified tool:

- Keeps the existing `advisor_recommendation_update` tool name, command ID, payload, response model, state validation, retry behavior, and error handling.
- Accepts `--subscription` or `--service-group` and rejects requests that explicitly provide both.
- Preserves configured-default subscription behavior when neither scope is supplied.
- Uses the existing `--recommendation-id`, status, postponement, dismissal-reason, and optional tenant options for both scopes.
- Calls the service-group ARM route:
  `PATCH /providers/Microsoft.Management/serviceGroups/{serviceGroupId}/providers/Microsoft.Advisor/recommendations/{recommendationId}`.
- Uses Advisor API version `2026-08-01-preview` for both subscription and service-group update routes.
- Uses the configured cloud's ARM endpoint and bearer-token audience.
- Does not perform subscription resolution or lookup for service-group requests.
- Uses the configured/default credential tenant when `--tenant` is omitted and resolves an explicit tenant when supplied.
- Routes subscription and service-group requests through the same request payload, authenticated Azure Core HTTP pipeline, response conversion, retry, and error-parsing code.

## Dependency and rollout

#3303 is merged. This branch now includes the resulting `upstream/main` commit, and the PR diff contains only the service-group enhancement over the merged subscription update tool.

The service-group PATCH endpoint is not yet available, so this PR must remain unmerged until the live PATCH acceptance matrix passes.

Public references:

- [Advisor Recommendations Update REST API](https://learn.microsoft.com/rest/api/advisor/recommendations/update?view=rest-advisor-2026-08-01-preview)
- [Azure Service Groups](https://learn.microsoft.com/azure/governance/service-groups/manage-service-groups)

## GitHub issue number?

Related to #3381 and follows merged PR #3303.

## Validation completed

- 394 Azure Advisor tests passed after merging current `upstream/main`.
- Service-group ID validation follows the official Service Groups contract: 1-250 characters, globally unique, with alphanumeric characters and `- _ ( ) . ~`; tests cover the real `guptaravi-sg-test` form, every special character, non-ASCII alphanumeric input, the 250-character boundary, and 251-character rejection.
- Advisor and Azure MCP Server builds completed with zero warnings and errors.
- `Build-Local.ps1 -VerifyNpx` passed; inner build and npm commands were inspected for failures.
- Repository spelling validation passed.
- Command metadata generation and README annotation validation passed.
- The SG changelog entry conforms to the documented schema. The full compiler is currently blocked by the unrelated `Bug Fixes` section in the `copilot-fix-ai-search-consolidated-description.yml` entry already present in `upstream/main`.
- Live read-only ARM authentication passed for:
  - explicit subscription with and without tenant;
  - configured-default subscription with and without tenant;
  - service-group scope with and without tenant.
- The dual-scope request is rejected before service invocation or authentication.
- ToolDescriptionEvaluator tested all 16 subscription and service-group update prompts, including completed/done, dismiss/ignore, postpone/snooze, reactivate/reopen/reset, and generic state-change language:
  - rank: **#1 for every prompt**;
  - score range: **0.4204–0.5441** after the final explanatory description refinement;
  - embedding endpoint preflight: HTTP 200 with 3,072 dimensions.
- The temporary Entra bearer-token evaluator patch was reversed and evaluator source files were verified clean before commit.
- NativeAOT was attempted with `Build-Local.ps1 -IncludeNative`. The inner native publish remains blocked by the existing Cosmos `Microsoft.Azure.Documents.JsonSerializable` IL2026/IL3050 baseline failure; no Advisor-specific NativeAOT warning or error was reported.

## Live test status

The Advisor test project does not have deterministic live-test infrastructure for provisioning actionable recommendations. The service-group PATCH endpoint is also not available yet.

Before merge, live validation will exercise the full pairwise transition matrix across `New`, `Postponed`, `Dismissed`, and `Completed`, including same-state idempotency, restoration to the original state, invalid scope/identifier cases, and every safely reproducible backend rejection.

## Pre-merge Checklist

- [x] Required for All PRs
    - [x] **Read [contribution guidelines](https://github.com/microsoft/mcp/blob/main/CONTRIBUTING.md)**
    - [x] PR title clearly describes the change
    - [x] Commit history is clean with descriptive messages ([cleanup guide](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md))
    - [x] Added comprehensive tests for new/modified functionality
    - [x] Created a changelog entry if the change falls among the following: new feature, bug fix, UI/UX update, breaking change, or updated dependencies. Follow the [changelog entry guide](https://github.com/microsoft/mcp/blob/main/docs/changelog-entries.md)
- [x] For MCP tool changes:
    - [x] **One tool per PR**: This PR modifies only `advisor_recommendation_update`
    - [x] Updated `servers/Azure.Mcp.Server/README.md`
    - [x] Validated README changes using `Process-PackageReadMe.ps1`
    - [x] Ran ToolDescriptionEvaluator and obtained score >= 0.4 and top-three ranking for every related prompt
    - [x] Updated `consolidated-tools.json`
    - [x] Added public Azure Advisor and Service Groups documentation links
- [x] Extra steps for **Azure MCP Server** tool changes:
    - [x] Updated `servers/Azure.Mcp.Server/docs/azmcp-commands.md`
    - [x] Ran `Update-AzCommandsMetadata.ps1`
    - [x] Updated `servers/Azure.Mcp.Server/docs/e2eTestPrompts.md`

## Invoking Livetests

Copilot submitted PRs are not trustworthy by default. Users with `write` access to the repo need to validate the contents of this PR before leaving a comment with the text `/azp run mcp - pullrequest - live`. This will trigger the necessary livetest workflows to complete required validation.

### CODE REVIEW

**Review by ankiga-MSFT — APPROVED** [review]

File: tools/Azure.Mcp.Tools.Advisor/src/Commands/Recommendation/RecommendationUpdateCommand.cs [hunk 1]
```diff
@@ -37,15 +35,51 @@ public sealed class RecommendationUpdateCommand(
...
+        var hasSubscription = !string.IsNullOrWhiteSpace(options.Subscription);
+        var hasServiceGroup = !string.IsNullOrWhiteSpace(options.ServiceGroup);
+        if (hasSubscription && hasServiceGroup)
+        {
+            validationResult.Errors.Add("Specify either --subscription or --service-group, not both.");
+        }
+        else if (!hasSubscription && !hasServiceGroup)
+        {
+            validationResult.Errors.Add("Missing Required options: --subscription or --service-group.");
+        }
```
[the remark this answers is not in the document]
Comment (ravigupta99) [reply tools/Azure.Mcp.Tools.Advisor/src/Commands/Recommendation/RecommendationUpdateCommand.cs]: The concern is valid and is fixed in fde083a06. Mutual exclusion now checks option presence with `is not null`, while `IsNullOrWhiteSpace` remains responsible for effective missing-scope validation. This rejects an explicitly supplied blank value on either side instead of routing through the other scope. I added regression coverage for `--subscription sub1 --service-group " "` and `--subscription " " --service-group sg1`, verifying HTTP 400, no default-subscription resolution, and no service call. All 378 Advisor tests pass and the Azure MCP Server builds with zero warnings.

File: tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ namespace Azure.Mcp.Tools.Advisor.Services;
 public class AdvisorService(IAzureService azureService)
     : BaseAzureResourceService(azureService), IAdvisorService
 {
-    private const string RecommendationUpdateApiVersion = "2026-03-01-preview";
+    private const string RecommendationUpdateApiVersion = "2026-08-01-preview";
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs:21]: This version is also used by the existing subscription update path. Since `2026-08-01-preview` is not currently published or listed in provider metadata, this could regress subscription updates. Could we keep subscriptions on `2026-03-01-preview` until the new version passes live validation for both scopes?

File: tools/Azure.Mcp.Tools.Advisor/src/Commands/Recommendation/RecommendationUpdateCommand.cs [hunk 3]
```diff
@@ -37,15 +35,52 @@ public sealed class RecommendationUpdateCommand(
...
+        else if (string.IsNullOrWhiteSpace(options.Subscription) &&
+            string.IsNullOrWhiteSpace(options.ServiceGroup))
+        {
+            validationResult.Errors.Add("Missing Required options: --subscription or --service-group.");
+        }
+
+        if (serviceGroupWasProvided &&
+            (options.ServiceGroup!.Length is < 1 or > 90 ||
+             !options.ServiceGroup.All(IsValidServiceGroupNameCharacter)))
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Advisor/src/Commands/Recommendation/RecommendationUpdateCommand.cs]: This validation is stricter than the Service Groups contract. Service-group IDs can be up to 250 characters and may contain `~`, so valid groups would be rejected here. Please align this check with the documented constraints.

File: tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs (hunk 2 above)
Comment (ravigupta99) [reply tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs:21]: We are not merging this PR as of yet, this PR will only be merged post 15th Sept when the service group support will also be lighted up from our side in our ARM APIs and ARG. And yes, at that time we will be releasing a new sweeping version of all our ARM APIs which will have patch support for both subscription pivoted and SG pivoted recommendations.

I will update the version at that time, if needed, but as of now the plan from Advisor side is to publish the 2026-08-01-preview version in our swagger spec and manifest in a couple of weeks with this support.

File: tools/Azure.Mcp.Tools.Advisor/src/Commands/Recommendation/RecommendationUpdateCommand.cs (hunk 3 above)
Comment (ravigupta99) [reply tools/Azure.Mcp.Tools.Advisor/src/Commands/Recommendation/RecommendationUpdateCommand.cs]: The concern is valid and is fixed in d338f18cf. The validator now follows the official Service Groups overview: IDs support up to 250 characters and alphanumeric characters plus `- _ ( ) . ~`. I also removed the undocumented ASCII-only restriction by using `char.IsLetterOrDigit`.

The updated tests cover a one-character ID, the real `guptaravi-sg-test` form, all documented special characters including `~`, non-ASCII alphanumeric input, exactly 250 characters, 251-character rejection, and undocumented `/` and `#` characters. All 394 Advisor tests pass, the Azure MCP Server builds successfully, VerifyNpx passes, and all 16 update prompts still rank #1 with scores 0.4255-0.5923.

Official contract: https://learn.microsoft.com/azure/governance/service-groups/overview#how-it-works

File: tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs (hunk 2 above)
Comment (ravigupta99) [reply tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs]: Updated status: I have kept versions for both different.

File: tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs [hunk 4]
```diff
@@ -18,7 +18,8 @@ namespace Azure.Mcp.Tools.Advisor.Services;
 public class AdvisorService(IAzureService azureService)
     : BaseAzureResourceService(azureService), IAdvisorService
 {
-    private const string RecommendationUpdateApiVersion = "2026-03-01-preview";
+    private const string SubscriptionRecommendationUpdateApiVersion = "2026-03-01-preview";
+    private const string ServiceGroupRecommendationUpdateApiVersion = "2026-08-01-preview";
```
Comment (g2vinay) [inline tools/Azure.Mcp.Tools.Advisor/src/Services/AdvisorService.cs:22]: Can you confirm this exact `2026-08-01-preview` API works on `management.azure.com`? The PR only shows canary testing with `2026-08-01-alpha`, and I could not find the service-group route in the public REST specs.

File: servers/Azure.Mcp.Server/docs/e2eTestPrompts.md [hunk 5]
```diff
@@ -91,9 +91,21 @@ The `Interaction` column describes whether a prompt can invoke its tool immediat
...
+| advisor_recommendation_update | Change Advisor recommendation <recommendation-id> state to Completed for service group <service-group> | none |
+| advisor_recommendation_update | Mark Advisor recommendation <recommendation-id> done for service group <service-group> | none |
 | advisor_recommendation_update | Dismiss Advisor recommendation <recommendation-id> because the risk is acceptable in subscription <subscription> | none |
+| advisor_recommendation_update | Dismiss Advisor recommendation <recommendation-id> because the risk is acceptable in service group <service-group> | none |
+| advisor_recommendation_update | Ignore Advisor recommendation <recommendation-id> because the risk is acceptable for service group <service-group> | none |
+| advisor_recommendation_update | Set Advisor recommendation <recommendation-id> status to Dismissed in subscription <subscription> | none |
 | advisor_recommendation_update | Postpone Advisor recommendation <recommendation-id> until December 31, 2026 in subscription <subscription> | none |
+| advisor_recommendation_update | Postpone Advisor recommendation <recommendation-id> until December 31, 2026 in service group <service-group> | none |
```
Comment (g2vinay) [inline servers/Azure.Mcp.Server/docs/e2eTestPrompts.md:102]: Please replace this fixed date with a placeholder such as `<future-date-time-with-offset>`. This prompt will fail once the date has passed. Please update the matching README example too.
